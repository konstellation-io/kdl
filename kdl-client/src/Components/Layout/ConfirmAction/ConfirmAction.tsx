import { ModalContainer, ModalLayoutInfo, ModalLayoutJustify } from 'kwc';
import React, { FC, useCallback, useEffect } from 'react';

import useBoolState from 'Hooks/useBoolState';
import { useForm } from 'react-hook-form';

type Props = {
  title: string;
  action: (msg: string) => void;
  actionLabel?: string;
  subtitle?: string;
  warning?: boolean;
  error?: boolean;
  message?: string;
  showInput?: boolean;
  confirmationWord?: string;
  children: JSX.Element;
};

type FormData = {
  message: string;
};

const ConfirmAction: FC<Props> = ({
  title,
  subtitle,
  action,
  children,
  message,
  confirmationWord,
  actionLabel,
  showInput = false,
  warning = false,
  error = false,
}) => {
  const {
    handleSubmit,
    setValue,
    register,
    unregister,
    errors,
    clearErrors,
  } = useForm<FormData>({
    defaultValues: {
      message: '',
    },
  });

  const {
    value: modalVisible,
    activate: showModal,
    deactivate: hideModal,
  } = useBoolState(false);

  const validateMessage = useCallback(
    (value: string) =>
      (confirmationWord &&
        confirmationWord !== value &&
        `You need to type: "${confirmationWord}"`) ||
      true,
    [confirmationWord]
  );

  useEffect(() => {
    register('message', {
      required: showInput && 'This field is mandatory',
      validate: validateMessage,
    });

    return () => unregister('message');
  }, [register, unregister, setValue, validateMessage, showInput]);

  function onSubmit({ message }: FormData) {
    hideModal();
    action(message);
  }

  return (
    <>
      <div
        onClick={(e) => {
          e.stopPropagation();
          showModal();
        }}
      >
        {children}
      </div>
      {modalVisible && (
        <ModalContainer
          title={title}
          subtitle={subtitle}
          onAccept={handleSubmit(onSubmit)}
          actionButtonLabel={actionLabel}
          onCancel={hideModal}
          warning={warning}
          error={error}
          blocking
        >
          {message && <ModalLayoutInfo>{message}</ModalLayoutInfo>}
          {showInput && (
            <ModalLayoutJustify
              onUpdate={(value: string) => {
                setValue('message', value);
                clearErrors();
              }}
              submit={handleSubmit(onSubmit)}
              error={errors?.message?.message || ''}
              label={
                confirmationWord
                  ? `WRITE "${confirmationWord}"`
                  : 'Why are you doing that?'
              }
              isInput={!!confirmationWord}
            />
          )}
        </ModalContainer>
      )}
    </>
  );
};

export default ConfirmAction;
