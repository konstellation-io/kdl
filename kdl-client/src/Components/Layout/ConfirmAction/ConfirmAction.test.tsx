import { ModalContainer, ModalLayoutJustify } from 'kwc';

import ConfirmAction from './ConfirmAction';
import React from 'react';
import { shallow } from 'enzyme';

const actionMock = jest.fn();

let component;

describe('ConfirmAction component - without input', () => {
  beforeEach(() => {
    component = shallow(
      <ConfirmAction
        title="Some title"
        action={actionMock}
        actionLabel="Action label"
        subtitle="Some subtitle"
        warning={false}
        error={false}
        message="Some message"
        showInput={false}
        confirmationWord={undefined}
      >
        <div id="child-1">Child 1</div>
      </ConfirmAction>
    );
  });

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    // First, make the modal visible
    component
      .find('div')
      .first()
      .simulate('click', { stopPropagation() {} });

    expect(component.contains('Some message')).toBeTruthy();
  });

  it('handles events', () => {
    // First, make the modal visible
    component
      .find('div')
      .first()
      .simulate('click', { stopPropagation() {} });

    // Then click on Accept button
    component.find(ModalContainer).props().onAccept();

    expect(actionMock).toHaveBeenCalledTimes(1);
  });
});

describe('ConfirmAction component - with input', () => {
  beforeEach(() => {
    component = shallow(
      <ConfirmAction
        title="Some title"
        action={actionMock}
        actionLabel="Action label"
        subtitle="Some subtitle"
        warning={false}
        error={false}
        message="Some message"
        showInput={true}
        confirmationWord={'WORD'}
      >
        <div id="child-1">Child 1</div>
      </ConfirmAction>
    );

    actionMock.mockClear();
  });

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    // First, make the modal visible
    component
      .find('div')
      .first()
      .simulate('click', { stopPropagation() {} });

    expect(component.contains('Some message')).toBeTruthy();
    expect(component.find(ModalLayoutJustify).props().label).toBe(
      'WRITE "WORD"'
    );
  });
});
