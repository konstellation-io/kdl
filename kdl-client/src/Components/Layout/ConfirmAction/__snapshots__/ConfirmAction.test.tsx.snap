// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmAction component - with input Component match snapshot 1`] = `
<Fragment>
  <div
    onClick={[Function]}
  >
    <div
      id="child-1"
    >
      Child 1
    </div>
  </div>
</Fragment>
`;

exports[`ConfirmAction component - without input Component match snapshot 1`] = `
<Fragment>
  <div
    onClick={[Function]}
  >
    <div
      id="child-1"
    >
      Child 1
    </div>
  </div>
</Fragment>
`;
